### ISSUE TITLE
[YCQL] Incorrect index update if used expression result is still unchanged.

### ISSUE THREAD

**Issue by OlegLoginov** [yugabyte/yugabyte-db#5898]
Reproducer:
```
CREATE TABLE test_update (h int, r int, c int, j JSONB, PRIMARY KEY ((h), r)) WITH transactions = {'enabled' : true};
CREATE INDEX i2 on test_update ((j->'a'->>'b'));
insert into test_update (h, r, c, j) values (1, 2, 3, '{\"a\":1}');
select * from i2;   <<<<<<<<< OK: Row[1, 2, NULL]

update test_update set j = '{\"a\":2}' where h = 1 and r = 2;
select * from i2;   <<<<<<<<< ERROR: Expected - Row[1, 2, NULL], but was []
```
The bug is in the index updating implementation for indexed column with expression.
Root cause: in the QLWriteOperation::UpdateIndexes() code we compare only column values instead of expression results on the column values.
Finally the same index PK was inserted and then immediately deleted. So, got the empty row set in the index (as in the example above).
From the source code point of view the bug is in incorrect calculated value for flag `index_key_changed`:
```
        if (new_row.IsColumnSpecified(index_column.indexed_column_id)) {
          RETURN_NOT_OK(expr_executor->EvalExpr(index_column.colexpr, new_row, result.Writer()));
          if (!new_row.MatchColumn(index_column.indexed_column_id, existing_row)) {     <<<<<<<<<<< new_col_val != exist_col_val, instead of EXPR(new_col_val) != EXPR(exist_col_val)
            index_key_changed = true;
          }
```
In the example `index_key_changed` must be FALSE, but it's TRUE by mistake. Finally the code does 2 write requests: 1 INSERT request + 1 DELETE request, as it supposes that the PK is changed.
But the PK is NOT changed in fact. So these 2 requests INSERT and then DELETE the same index row (with same PK).
```
      QLWriteRequestPB* index_request =
          (index->HasWritePermission() ? NewIndexRequest(index,
                                                         QLWriteRequestPB::QL_STMT_INSERT,
                                                         new_row)
                                       : nullptr);                                         <<< INSERT REQUEST
      RETURN_NOT_OK(PrepareIndexWriteAndCheckIfIndexKeyChanged(
          this, existing_row, new_row, index, index_request, &index_key_changed));
    }

    // If the index key is changed, delete the current key.
    if (index_key_changed && index->HasDeletePermission()) {
      QLWriteRequestPB* index_request =
          NewIndexRequest(index, QLWriteRequestPB::QL_STMT_DELETE, new_row);   <<< DELETE REQUEST for the same PK due to invalid 'index_key_changed' == TRUE
```

**Comment by OlegLoginov** [comment]
Affected indexes:
1. Secondary indexes with an EXPRESSION over a PK (in the index) column.
2. If an INSERT/UPDATE statement changes the indexed column value, BUT the EXPRESSION values is NOT changed.

**Comment by OlegLoginov** [comment]
WAS:
```
          RETURN_NOT_OK(expr_executor->EvalExpr(index_column.colexpr, new_row, result.Writer()));
          if (!new_row.MatchColumn(index_column.indexed_column_id, existing_row)) {     <<<<<<<<<<< new_col_val != exist_col_val, instead of EXPR(new_col_val) != EXPR(exist_col_val)
            index_key_changed = true;
          }
```
FIX:
```
          RETURN_NOT_OK(expr_executor->EvalExpr(index_column.colexpr, new_row, result.Writer()));
          // Compare new and existing results of the expression, if the results are equal
          // that means the key is NOT changed in fact even if the column value is changed.
          QLExprResult existing_result;
          RETURN_NOT_OK(expr_executor->EvalExpr(index_column.colexpr, existing_row, existing_result.Writer()));
          if (result.Value() != existing_result.Value()) {                <<<<<<<<<<< <<<<<< COMPARE only  EXPR(column values)
            index_key_changed = true;
          }
```

**Comment by OlegLoginov** [comment]
Fixed & backported into releases 2.1 & 2.2.
